Crash under WTF::charactersAreAllASCII<unsigned char>(unsigned char const*, unsigned long)

cdumez opened the pull request:
#### 147e4eda9ad90bbd17194b948d7b35c477ffddad
<pre>
Crash under WTF::charactersAreAllASCII&lt;unsigned char&gt;(unsigned char const*, unsigned long)
<a href="https://bugs.webkit.org/show_bug.cgi?id=242899">https://bugs.webkit.org/show_bug.cgi?id=242899</a>
&lt;rdar://94931119&gt;

Reviewed by Cameron McCormack.

We&apos;re crashing when reading the memory of a service worker script. Such memory is often mmap&apos;d
from disk to reduce memory usage. Such crash could theoretically occur if the file gets modified
on disk so that its size becomes smaller, *after* it&apos;s been mmap&apos;d and *before* the mmap&apos;d memory
is read. Normally, we protect against this by deleting the existing script file on disk before
attempting to write it, as this keeps existing mmap&apos;d handles valid. However, there was one code
path inside SWScriptStorage::store() where we were failing to delete the file before attempting
to write it. Namely, in the case where the new size of the file is less than the page size, we
would use a different code path where we choose not to mmap(), and we would fail to delete the
existing file then. This file addresses this issue.

This is a speculative fix for the crash in the radar which I believe could happen if:
1. Write service worker script file with size &gt; pageSize to disk and mmap it
2. Rewrite service worker script file with size &lt; pageSize to disk
3. Attempt to read the mmap&apos;d memory from step 1

* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::store):

Canonical link: <a href="https://commits.webkit.org/252617@main">https://commits.webkit.org/252617@main</a>
</pre>

heycam reviewed (approved):
Your speculation sounds good!

cdumez committed:
Crash under WTF::charactersAreAllASCII<unsigned char>(unsigned char const*, unsigned long)
https://bugs.webkit.org/show_bug.cgi?id=242899
<rdar://94931119>

Reviewed by Cameron McCormack.

We're crashing when reading the memory of a service worker script. Such memory is often mmap'd
from disk to reduce memory usage. Such crash could theoretically occur if the file gets modified
on disk so that its size becomes smaller, *after* it's been mmap'd and *before* the mmap'd memory
is read. Normally, we protect against this by deleting the existing script file on disk before
attempting to write it, as this keeps existing mmap'd handles valid. However, there was one code
path inside SWScriptStorage::store() where we were failing to delete the file before attempting
to write it. Namely, in the case where the new size of the file is less than the page size, we
would use a different code path where we choose not to mmap(), and we would fail to delete the
existing file then. This file addresses this issue.

This is a speculative fix for the crash in the radar which I believe could happen if:
1. Write service worker script file with size > pageSize to disk and mmap it
2. Rewrite service worker script file with size < pageSize to disk
3. Attempt to read the mmap'd memory from step 1

* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::store):

Canonical link: https://commits.webkit.org/252617@main
